src/Bicep.Cli/Logging/BicepDiagnosticLogger.cs

```diff
@@ -29,20 +38,123 @@ public void LogDiagnostic(Uri fileUri, IDiagnostic diagnostic, ImmutableArray<in
-
-            this.logger.Log(ToLogLevel(diagnostic.Level), message);
+            if (Format == DiagnosticsFormat.Default) 
+            {
+                var message = $"{fileUri.LocalPath}({line + 1},{character + 1}) : {diagnostic.Level} {diagnostic.Code}: {diagnostic.Message}{codeDescription}";
+                this.logger.Log(ToLogLevel(diagnostic.Level), message);
+            }
+            else if (Format == DiagnosticsFormat.Json)
+            {
+                sarifResults.Add(new Result()
+                {
+                    RuleId = diagnostic.Code,
+                    Level = ToFailureLevel(diagnostic.Level),
+                    Message = new Message
+                    {
+                        Text = $"{diagnostic.Message}{codeDescription}",
+                    },
+                    Locations = new[]
+                    {
+                        new Location
+                        {
+                            PhysicalLocation = new PhysicalLocation
+                            {
+                                ArtifactLocation = new ArtifactLocation
+                                {
+                                    Uri = fileUri
+                                },
+                                Region = new Region
+                                {
+                                    StartLine = line + 1,
+                                    CharOffset = character + 1,
+                                }
+                            }
+                        }
+                    }
+                });
+            }
 
             // Increment counters
             if (diagnostic.Level == DiagnosticLevel.Warning) { this.WarningCount++; }
             if (diagnostic.Level == DiagnosticLevel.Error) { this.ErrorCount++; }
         }
 
+
+        private enum DiagnosticsFormat
+        {
+            Default,
+            Json
+        }
+
         public int ErrorCount { get; private set; }
 
         private int WarningCount { get; set; }
 
+        private DiagnosticsFormat Format { get; set; }
+
+        private SarifLog sarifLog { get; set; }
+
+        private List<Result> sarifResults { get; set; }
+
+
+        public void SetupFormat(string? format)
+        {
+            Format = ToDiagnosticsFormat(format);
+            if(Format == DiagnosticsFormat.Json)
+            {
+                // setup an empty run on the sarif log to add results to.
+                sarifLog.Runs = new List<Run>();
+                sarifLog.Runs.Add(new Run()
+                {
+                    Tool = new Tool(new ToolComponent()
+                    {
+                        Name = "bicep",
+                    }, null, null)
+                });
+            }
+        }
+
+        public void FlushLog()
+        {
+            if(Format == DiagnosticsFormat.Json && sarifResults.Count > 0)
```

Comment:
>sarifResults.Count > 0 [](http://example.com/codeflow?start=51&length=22)

What would be the downside of emitting a Sarif file with 0 diagnostics in it? (Is that allowed at all?)

Reply:
It is allowed (an example would be [this](https://github.com/microsoft/sarif-tutorials/blob/main/samples/2-Basics/empty-log.sarif.json)) ! I think I was going off how we output nothing in the scenario where we don't have any errors.

Reply:
I was thinking of slightly simplifying automation scenarios. If we emit a Sarif file with zero diagnostics, there are fewer cases for the caller to deal with as you get a Sarif file no matter the result.

Reply:
Updated to emit a sarif file regardless of if there are results.